HDDS-1913. Fix OzoneBucket and RpcClient APIS for acl.

bharatviswa504 opened the pull request:
(cherry picked from commit b60d93b8b500f0ba97c027125401368d22028822)

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 85 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 8 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 26 | Maven dependency ordering for branch |
| +1 | mvninstall | 642 | trunk passed |
| +1 | compile | 391 | trunk passed |
| +1 | checkstyle | 77 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 1059 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 238 | trunk passed |
| 0 | spotbugs | 539 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 807 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 31 | Maven dependency ordering for patch |
| +1 | mvninstall | 655 | the patch passed |
| +1 | compile | 453 | the patch passed |
| +1 | cc | 453 | the patch passed |
| +1 | javac | 453 | the patch passed |
| +1 | checkstyle | 109 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 868 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 205 | the patch passed |
| -1 | findbugs | 551 | hadoop-ozone generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0) |
||| _ Other Tests _ |
| -1 | unit | 395 | hadoop-hdds in the patch failed. |
| -1 | unit | 3658 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 61 | The patch does not generate ASF License warnings. |
| | | 10802 | |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-ozone |
|  |  Unread field:OzoneBucket.java:[line 145] |
|  |  Unwritten field:BucketArgs.java:[line 88] |
| Failed junit tests | hadoop.hdds.scm.block.TestBlockManager |
|   | hadoop.ozone.client.rpc.TestMultiBlockWritesWithDnFailures |
|   | hadoop.ozone.client.rpc.TestContainerStateMachineFailures |
|   | hadoop.ozone.client.rpc.TestOzoneAtRestEncryption |
|   | hadoop.ozone.om.TestScmSafeMode |
|   | hadoop.ozone.client.rpc.Test2WayCommitInRatis |
|   | hadoop.ozone.client.rpc.TestCommitWatcher |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineCreateAndDestory |
|   | hadoop.ozone.client.rpc.TestBlockOutputStream |
|   | hadoop.ozone.om.TestKeyManagerImpl |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.0 Server=19.03.0 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1257 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc |
| uname | Linux 3b2f9bdfcadc 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 5840df8 |
| Default Java | 1.8.0_212 |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/1/artifact/out/new-findbugs-hadoop-ozone.html |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/1/testReport/ |
| Max. process+thread count | 3877 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/common hadoop-ozone/client hadoop-ozone/ozone-manager hadoop-ozone/objectstore-service hadoop-ozone/integration-test U: hadoop-ozone |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

xiaoyuyao commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/BucketManagerImpl.java:
```diff
@@ -272,16 +272,6 @@ public void setBucketProperty(OmBucketArgs args) throws IOException {
-        LOG.debug("Updating ACLs for bucket: {} in volume: {}",
-            bucketName, volumeName);
-      } else {
-        bucketInfoBuilder.setAcls(oldBucketInfo.getAcls());
```
We need to keep bucketInfoBuilder.setAcls(oldBucketInfo.getAcls()) otherwise existing acls will be lost upon setBucketProperty. Can you add a unit test to cover this?

bharatviswa504 replied:
Thanks for catching fixed it, and also added a test to cover this scenario.

bharatviswa504 commented:
Thank You @xiaoyuyao for the review.
Addressed review comments, and also fixed another issue when creating bucket in one API, not setting default ACL list.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 44 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 10 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 31 | Maven dependency ordering for branch |
| +1 | mvninstall | 591 | trunk passed |
| +1 | compile | 350 | trunk passed |
| +1 | checkstyle | 61 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 794 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 152 | trunk passed |
| 0 | spotbugs | 413 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 602 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 21 | Maven dependency ordering for patch |
| +1 | mvninstall | 533 | the patch passed |
| +1 | compile | 357 | the patch passed |
| +1 | cc | 357 | the patch passed |
| +1 | javac | 357 | the patch passed |
| +1 | checkstyle | 68 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 648 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 148 | the patch passed |
| +1 | findbugs | 678 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 288 | hadoop-hdds in the patch passed. |
| -1 | unit | 1261 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 38 | The patch does not generate ASF License warnings. |
| | | 6832 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdds.scm.pipeline.TestNodeFailure |
|   | hadoop.ozone.om.TestKeyManagerImpl |
|   | hadoop.ozone.TestMiniOzoneCluster |
|   | hadoop.ozone.web.TestOzoneWebAccess |
|   | hadoop.ozone.web.TestOzoneVolumes |
|   | hadoop.ozone.web.client.TestBuckets |
|   | hadoop.ozone.web.client.TestVolume |
|   | hadoop.ozone.om.TestScmSafeMode |
|   | hadoop.ozone.container.common.statemachine.commandhandler.TestCloseContainerHandler |
|   | hadoop.ozone.web.client.TestKeysRatis |
|   | hadoop.ozone.web.TestOzoneRestWithMiniCluster |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1257 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc |
| uname | Linux 01a93ff1f685 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 6ff0453 |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/2/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/2/testReport/ |
| Max. process+thread count | 3625 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/common hadoop-ozone/client hadoop-ozone/ozone-manager hadoop-ozone/objectstore-service hadoop-ozone/s3gateway hadoop-ozone/integration-test U: hadoop-ozone |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

xiaoyuyao commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:
```diff
@@ -388,7 +388,9 @@ public void deleteVolume(String volumeName) throws IOException {
-    createBucket(volumeName, bucketName, BucketArgs.newBuilder().build());
+    // Set acls of current user.
+    createBucket(volumeName, bucketName,
+        BucketArgs.newBuilder().setAcls(getAclList()).build());
```
This will cause creator acls added twice into the final list because the same list will be added again when the passin bucketArgs has a non-empty acl list. 
{code}
    List<OzoneAcl> listOfAcls = getAclList();
    //ACLs from BucketArgs
    if(bucketArgs.getAcls() != null) {
      listOfAcls.addAll(bucketArgs.getAcls());
    }
{code}

The list of acls are not merged properly when using List#addAll, which will be fixed in HDDS-1927. 
Let's file a separate JIRA for RpcClient#createBucket issue.

bharatviswa504 replied:
Done.

xiaoyuyao commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/OMBucketSetPropertyRequest.java:
```diff
@@ -134,17 +132,6 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
-        LOG.debug("Updating ACLs for bucket: {} in volume: {}",
-            bucketName, volumeName);
-      } else {
-        bucketInfoBuilder.setAcls(oldBucketInfo.getAcls());
```
We have a similar problem here because OMBucketArgs does not have acls, we will need to rely on the oldBucketInfo.getAcls to avoid reset existing acl on bucket.

bucketInfoBuilder.setAcls(oldBucketInfo.getAcls());

bharatviswa504 commented:
/retest

xiaoyuyao commented:
+1, pending CI... Thanks for the quick fix.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 73 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 10 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 29 | Maven dependency ordering for branch |
| +1 | mvninstall | 602 | trunk passed |
| +1 | compile | 366 | trunk passed |
| +1 | checkstyle | 72 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 983 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 223 | trunk passed |
| 0 | spotbugs | 475 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 708 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 27 | Maven dependency ordering for patch |
| +1 | mvninstall | 643 | the patch passed |
| +1 | compile | 413 | the patch passed |
| +1 | cc | 413 | the patch passed |
| +1 | javac | 413 | the patch passed |
| +1 | checkstyle | 86 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 803 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 198 | the patch passed |
| +1 | findbugs | 707 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 356 | hadoop-hdds in the patch passed. |
| -1 | unit | 2145 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 42 | The patch does not generate ASF License warnings. |
| | | 8671 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestOzoneRpcClient |
|   | hadoop.ozone.client.rpc.TestMultiBlockWritesWithDnFailures |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |
|   | hadoop.ozone.client.rpc.TestOzoneAtRestEncryption |
|   | hadoop.ozone.container.common.statemachine.commandhandler.TestBlockDeletion |
|   | hadoop.ozone.client.rpc.TestSecureOzoneRpcClient |
|   | hadoop.ozone.om.TestKeyManagerImpl |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineCreateAndDestory |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1257 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc |
| uname | Linux 5dd0fbb2f379 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / ce3c5a3 |
| Default Java | 1.8.0_222 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/3/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/3/testReport/ |
| Max. process+thread count | 4909 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/common hadoop-ozone/client hadoop-ozone/ozone-manager hadoop-ozone/objectstore-service hadoop-ozone/s3gateway hadoop-ozone/integration-test U: hadoop-ozone |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 82 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 10 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 13 | Maven dependency ordering for branch |
| +1 | mvninstall | 616 | trunk passed |
| +1 | compile | 405 | trunk passed |
| +1 | checkstyle | 73 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 962 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 188 | trunk passed |
| 0 | spotbugs | 482 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 701 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 18 | Maven dependency ordering for patch |
| +1 | mvninstall | 592 | the patch passed |
| +1 | compile | 378 | the patch passed |
| +1 | cc | 378 | the patch passed |
| +1 | javac | 378 | the patch passed |
| +1 | checkstyle | 74 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 775 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 188 | the patch passed |
| +1 | findbugs | 739 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 372 | hadoop-hdds in the patch passed. |
| -1 | unit | 2293 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 48 | The patch does not generate ASF License warnings. |
| | | 8674 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.container.common.transport.server.ratis.TestCSMMetrics |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |
|   | hadoop.ozone.container.server.TestSecureContainerServer |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClientForAclAuditLog |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineCreateAndDestory |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.0 Server=19.03.0 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1257 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc |
| uname | Linux 22107f15db23 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 9b8359b |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/4/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/4/testReport/ |
| Max. process+thread count | 4558 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/common hadoop-ozone/client hadoop-ozone/ozone-manager hadoop-ozone/objectstore-service hadoop-ozone/s3gateway hadoop-ozone/integration-test U: hadoop-ozone |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1257/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

bharatviswa504 commented:
/retest

bharatviswa504 commented:
Test failures are not related to this patch.
I will commit this to the trunk.
